pcdshub/lcls-plc-kfe-rix-motion
# ENH: motion/cc API update, ppm+wfs pmps coupling

## Description

- From https://jira.slac.stanford.edu/browse/ECS-1046
- Implement PMPS/WFS pmps parameter coupling for the PF1K2/IM5K2 pair and the PF2K2/IM6K2 pair
- Update to the latest versions of the accompanying common component APIs as needed to implement this feature
- Take the time to "standardize" the special cases here to the extent possible- no more PLC-specific function blocks for attenuator motion or physics calculations.
- Try to minimize motion deprecated API warnings
- Limit the number of exported PMPS PVs for IOC load times
- pin TwinCAT version to 4022.30 to match the PLC version
- hide unused config areas in the TwinCAT VS IDE

TODO

- [x] Finish minimizing deprecated API warnings
- [x] Test on the beamline
- [x] Minimize state device sizes

Things to test per device:

- Check that it can move to each state
- Check that the PMPS GUI updates appropriately during moves
- For the WFS, check that they have lighter requirements with ppm coupling

Devices to test:

- [x] AL1K2 e cannot reach out (top limit is at -37.429, out is at -33.5) [known issue]
- [x] AT1K2 w blade 3 both lims hit (is blade 3 even real?)
- [x] AT2K2
- [x] IM1K1 note: changed speed to 1 for safety because I was not in the FEE
- [x] IM1K2
- [x] IM2K2 note: changed speed to 15 to avoid resonance
- [x] IM3K2 note: changed speed to 14 to avoid resonance
- [x] IM4K2
- [x] IM5K2
- [x] IM6K2 note: changed speed to 14 to avoid resonance
- [x] LI2K2_K2A w cannot reach t1 or m2, hits limit
- [x] LI3K2_K2B
- [x] PF1K2 w cannot reach t1, hits limit
- [x] PF2K2
- [x] TM1K2
- [x] TM2K2

## Motivation and Context


- If the WFS is in, the YAG can take more beam, and furthermore it needs to in order to get a good signal.
- I needed to extend the common component and state API in a backwards incompatible way in order to allow for some of the dynamic behavior here
- Slow IOC load times are killer

## How Has This Been Tested?



I built the IOC and analyzed the diff to make sure no PVs were unintentionally removed.
Many PVs were intentionally removed.
I double-checked the ffo PV counts
I moved all of the state devices to all/most of their states
(this revealed some typos that I needed to fix)

## Where Has This Been Documented?


Here only

## Pre-merge checklist
- [x] Code contains descriptive comments
- [x] Libraries are set to fixed versions and not ``Always Newest``
- [x] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)
